lld/ELF/NanoMipsTransformationProperties.td

```diff
@@ -0,0 +1,655 @@
+
+class Arch;
+
+def NanoMips: Arch;
+
+// Properties of relocations
+class RelocProperty<int instSize, int bitsToRelocate, bits<64> insMask>{
+    // Size of instruction this relocation refers to in bits     
+    int InstSize = instSize;
+    // Num of bits that this relocation relocates (alters)     
+    int BitsToRelocate = bitsToRelocate;
+    // Mask for recognizing instructions that can be used with this relocation
+    // Used to find instruction property of the given instruction  
+    bits<64> InsMask = insMask;
+}
+
+// 16-bit reloc property
+class RelocProperty16<int bitsToRelocate, bits<64> insMask> : RelocProperty<16, bitsToRelocate, insMask>;
+
+
+// 16-bit reloc property with highest 6 bits set in mask 
+class RelocProperty16HI6<int bitsToRelocate> : RelocProperty16<bitsToRelocate, 0xfc00>;
+
+// 32-bit reloc property
+class RelocProperty32<int bitsToRelocate, bits<64> insMask> : RelocProperty<32, bitsToRelocate, insMask>;
+
+// 32-bit reloc property without a mask
+class RelocProperty32NoMask<int bitsToRelocate> : RelocProperty32<bitsToRelocate, 0>;
+
+// 48-bit reloc property
+class RelocProperty48<int bitsToRelocate, bits<64> insMask> : RelocProperty<48, bitsToRelocate, insMask>;
+
+// 48 bit instructions are treated as 16 bit as their opcode and regs are always in the first 16 bit, the rest 32 are for offset
+// and/or immediate
+
+// 48 bit reloc property with 6 highest bit and lowest 5 bits in mask set
+class RelocProperty48HI6LO5<int bitsToRelocate> : RelocProperty48<bitsToRelocate, 0xfc1f>;
+
+// ------------------------------- RelocProperty -------------------------------
+
+def NONE : RelocProperty<0, 0, 0>;
+def ALIGN : RelocProperty<0, 0, 0>;
+def PC4_S1 : RelocProperty16HI6<5>;
+def PC14_S1 : RelocProperty32<15, 0xfc00c000>;
+def PC21_S1 : RelocProperty32<22, 0xfc000000>;
+def PC_I32 : RelocProperty48HI6LO5<32>;
+def GPREL19_S2: RelocProperty32<21, 0xfc000003>;
+def GPREL18: RelocProperty32<18, 0xfc1c0000>;
+def GPREL17_S1: RelocProperty32<18, 0xfc1c0001>;
+def PC10_S1: RelocProperty16HI6<11>;
```

Comment:
Should R_NANOMIPS_GPREL16_S2 (defined in llvm/include/llvm/BinaryFormat/ELFRelocs/NanoMips.def) define its RelocProperty here like other similar (GPREL) relocation types? 
(There seems to be more also missing their definitions, like GPREL18_S3, PC32)

Reply:
Well, in general case you are right. But, for now these relocation properties are tied to the linker transformations and since not all relocations are included in transformations those relocations are missing from this file. If a new relocation is used in transformations, than the relocation should be definitely added here. Also, if RelocProperties find another usage other than transforming, other relocs should be added here. 

So currently, there are no transformations for R_NANOMIPS_GPREL16_S2 in this implementation, so it isn't present in this file.

Reply:
Using the same example, could you elaborate a little bit further on why R_NANOMIPS_GPREL17_S1 is included in transformations but R_NANOMIPS_GPREL16_S2 isn't? Does it also mean that R_NANOMIPS_GPREL16_S2 is currently not generated in object files by assembler? From the document (https://codescape.mips.com/components/toolchain/nanomips/2019.03-01/docs/MIPS_nanoMIPS_ABI_supplement_01_03_DN00179.pdf), looks like both relocation types are used in some instructions.

Reply:
Well, there are some reasons why R_NANOMIPS_GPREL16_S2 (as well as R_NANOMIPS_GPREL18_S3) is not included as a RelocProperty. The primary one is that it is not included in gold transformations, which was kind of the base for lld.
The second one is that I didn't manage to generate any of them as the instructions that use these relocations aren't mentioned in the ISA document (https://s3-eu-west-1.amazonaws.com/downloads-mips/I7200/I7200+product+launch/MIPS_nanomips32_ISA_TRM_01_01_MD01247.pdf).

Opposite to these two, there are transformations for R_NANOMIPS_GPREL17_S1 in gold and I managed to generate it through instructions that use this relocation. 

Anyways, I think I need to investigate this a little bit more, as this is odd. I will see whether it is possible to generate these instructions and also whether they should be relaxed/expanded. Thank you!

As for R_NANOMIPS_PC32, this is not in RelocProperties, as it is used for data relocations and it is not documented in the ABI supplement. I think it is used in .eh_frame and debug section relocations. The data in these sections won't be expanded/relaxed only the instruction sections.